components/script/dom/gamepadlist.rs

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::cell::DOMRefCell;
+use dom::bindings::codegen::Bindings::GamepadListBinding;
+use dom::bindings::codegen::Bindings::GamepadListBinding::GamepadListMethods;
+use dom::bindings::js::{JS, Root};
+use dom::bindings::reflector::{Reflector, reflect_dom_object};
+use dom::gamepad::Gamepad;
+use dom::globalscope::GlobalScope;
+use dom_struct::dom_struct;
+
+// https://www.w3.org/TR/gamepad/
+#[dom_struct]
+pub struct GamepadList {
+    reflector_: Reflector,
+    list: DOMRefCell<Vec<JS<Gamepad>>>
+}
+
+impl GamepadList {
+    fn new_inherited(list: &[&Gamepad]) -> GamepadList {
+        GamepadList {
+            reflector_: Reflector::new(),
+            list: DOMRefCell::new(list.iter().map(|g| JS::from_ref(&**g)).collect())
+        }
+    }
+
+    pub fn new(global: &GlobalScope, list: &[&Gamepad]) -> Root<GamepadList> {
+        reflect_dom_object(box GamepadList::new_inherited(list),
+                           global,
+                           GamepadListBinding::Wrap)
+    }
+
+    fn add(&self, gamepad: &Gamepad) {
+        self.list.borrow_mut().push(JS::from_ref(&*gamepad));
```

Comment:
This method was only called internally by add_if_not_exists. I have removed it to avoid confusions with the add_if_not_exists method.